test/common/buffer/buffer_fuzz_test.cc

```diff
@@ -0,0 +1,253 @@
+    ASSERT(previous_length - drain_length == target_buffer.length());
+    break;
+  }
+  case test::common::buffer::Action::kLinearize: {
+    const uint32_t linearize_size =
+        std::min(static_cast<uint32_t>(target_buffer.length()), action.linearize());
+    target_buffer.linearize(linearize_size);
+    Buffer::RawSlice slices[1];
+    const uint64_t slices_used = target_buffer.getRawSlices(slices, 1);
+    if (linearize_size > 0) {
+      ASSERT(slices_used == 1);
+      ASSERT(slices[0].len_ >= linearize_size);
+    }
+    break;
+  }
+  case test::common::buffer::Action::kMove: {
+    const uint32_t source_index = action.move().source_index() % BufferCount;
+    if (target_index == source_index) {
+      break;
+    }
+    Buffer::OwnedImpl& source_buffer = buffers[source_index];
+    if (action.move().length() == 0) {
+      target_buffer.move(source_buffer);
+    } else {
+      target_buffer.move(source_buffer, std::min(static_cast<uint32_t>(source_buffer.length()),
+                                                 action.move().length()));
+    }
+    break;
+  }
+  case test::common::buffer::Action::kRead: {
+    const uint32_t max_length = clampSize(action.read());
+    if (max_length == 0) {
+      break;
+    }
+    int pipe_fds[2] = {0, 0};
+    ASSERT(::pipe(pipe_fds) == 0);
+    ASSERT(::fcntl(pipe_fds[0], F_SETFL, O_NONBLOCK) == 0);
+    ASSERT(::fcntl(pipe_fds[1], F_SETFL, O_NONBLOCK) == 0);
+    std::string data(max_length, 'd');
+    const int rc = ::write(pipe_fds[1], data.data(), max_length);
+    ASSERT(rc > 0);
+    ASSERT(target_buffer.read(pipe_fds[0], max_length).rc_ == rc);
+    ASSERT(::close(pipe_fds[0]) == 0);
+    ASSERT(::close(pipe_fds[1]) == 0);
+    break;
+  }
+  case test::common::buffer::Action::kWrite: {
+    int pipe_fds[2] = {0, 0};
+    ASSERT(::pipe(pipe_fds) == 0);
+    ASSERT(::fcntl(pipe_fds[0], F_SETFL, O_NONBLOCK) == 0);
+    ASSERT(::fcntl(pipe_fds[1], F_SETFL, O_NONBLOCK) == 0);
+    const bool empty = target_buffer.length() == 0;
+    const int rc = target_buffer.write(pipe_fds[1]).rc_;
+    ASSERT(empty ? rc == 0 : rc > 0);
+    ASSERT(::close(pipe_fds[0]) == 0);
+    ASSERT(::close(pipe_fds[1]) == 0);
+    break;
+  }
+  default:
+    // Maybe nothing is set?
+    break;
+  }
+}
+
+} // namespace
+
+// Fuzz the owned buffer implementation.
+DEFINE_PROTO_FUZZER(const test::common::buffer::BufferFuzzTestCase& input) {
+  Context ctxt;
+  // Fuzzed buffers.
+  Buffer::OwnedImpl buffers[BufferCount];
+  // Shadows of buffers that are linearized at the end of each operation. This
+  // is an approximation of a simple buffer implementation without buffer
+  // slices.
+  Buffer::OwnedImpl linear_buffers[BufferCount];
+
+  for (int i = 0; i < input.actions().size(); ++i) {
+    const auto& action = input.actions(i);
+    ENVOY_LOG_MISC(debug, "Action {}", action.DebugString());
+    bufferAction(ctxt, buffers, action);
+    bufferAction(ctxt, linear_buffers, action);
```

Comment:
Question: Did you consider using something like `std::string` for the shadow implementation? It would obviously be a lot slower than our real buffer and more work, but a different implementation would show up issues more clearly. My concern is that we might have a bug that is present in both buffers, and isn't exposed when we linearize one of them. Did you consider this? I guess the the only operation that wouldn't pretty trivially map to `std::string` is the buffer fragments, so I guess we can't do it for that case, but I'm wondering if maybe we should do an `std::string` impl for fuzz cases that don't contain any buffer fragment operations? Just throwing this out there for thought, cc @brian-pane

Reply:
Yeah, I can do this. I was trying to be lazy and pretend that a normal buffer that was linearized after each operation would approximate this, but I don't mind doing the `std::string` implementation. I will specialize it just for `buffer_fuzz_test` and home it in `buffer_fuzz_test.cc`, so that we don't need to be in the business of writing validation tests for our validation test infrastructure.